### PR TITLE
all: Rework Opt value handling

### PR DESCRIPTION
**Pull request by fmeum**
This is a complete rewrite of the logic underlying `Opt`, with the
following notable changes:
* All options are now evaluated lazily, with safeguards ensuring that
  there can be no attempts to modify their defaults after they have been
  read once. This reduces the risk of accidentally locking in `Opt`
  and then reading stale values due to its previous reliance on static
  initializers.
* There are now clear precedence rules for all options and every option
  can be set from all possible sources: manifest entries, env variables,
  system properties, configuration parameters, command-line arguments.

In follow-up PRs, the new capabilities of `OptItem` (e.g. accumulating
settings and JUnit configuration parameters as sources) will be used in
Jazzer.

### CODE REVIEW

File: src/main/java/com/code_intelligence/jazzer/driver/Opt.java [hunk 1]
```diff
@@ -51,170 +42,125 @@ public final class Opt {
...
-          "target_args", ' ', "Arguments to pass to the fuzz target's fuzzerInitialize method")
-      : unmodifiableList(concat(Stream.of(autofuzz), autofuzzIgnore.stream()).collect(toList()));
+  // Internal options:
 
-  // Default to false if hooks is false to mimic the original behavior of the native fuzz target
-  // runner, but still support hooks = false && dedup = true.
-  public static final boolean dedup =
-      boolSetting("dedup", hooks, "Compute and print a deduplication token for every finding");
-
-  public static final String androidBootclassJarPath = stringSetting("android_bootclass_jar_path",
```
[the remark this answers is not in the document]
Comment (TheCoryBarker) [reply src/main/java/com/code_intelligence/jazzer/driver/Opt.java:152]: yes it's unused

File: src/main/java/com/code_intelligence/jazzer/junit/AgentConfigurator.java [hunk 2]
```diff
@@ -32,13 +35,12 @@ static void forRegressionTest(ExtensionContext extensionContext) {
     applyCommonConfiguration();
 
     // Add logic to the hook instrumentation that allows us to enable and disable hooks at runtime.
-    System.setProperty("jazzer.internal.conditional_hooks", "true");
+    Opt.conditionalHooks.setIfDefault(true);
     // Apply all hooks, but no coverage or compare instrumentation.
-    System.setProperty("jazzer.instrumentation_excludes", "**");
+    Opt.instrumentationIncludes.setIfDefault(singletonList("**"));
```
Comment (bertschneider) [inline src/main/java/com/code_intelligence/jazzer/junit/AgentConfigurator.java:40]: Why set `instrumentationIncludes` here?

File: src/main/java/com/code_intelligence/jazzer/driver/OptItem.java [hunk 3]
```diff
@@ -0,0 +1,383 @@
...
+      return Optional.of(splitOnUnescapedSeparator(rawValue, separator));
+    }
+
+    @Override
+    protected String getType() {
+      return String.format("list separated by '%s'", separator);
+    }
+
+    /**
+     * Split value into non-empty takens separated by separator. Backslashes can be used to escape
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply src/main/java/com/code_intelligence/jazzer/driver/OptItem.java]: Changed.

File: src/main/java/com/code_intelligence/jazzer/driver/OptItem.java [hunk 4]
```diff
@@ -0,0 +1,383 @@
...
+    Stream.Builder<String> builder = Stream.<String>builder().add(ROOT_SEGMENT);
+    if (isInternal()) {
+      builder.add(INTERNAL_SEGMENT);
+    }
+    builder.add(name);
+    return builder.build();
+  }
+
+  public static class IllegalOptionValueException extends IllegalArgumentException {
+    protected <T> IllegalOptionValueException(String message) {
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply src/main/java/com/code_intelligence/jazzer/driver/OptItem.java]: Removed.

File: src/main/java/com/code_intelligence/jazzer/junit/AgentConfigurator.java (hunk 2 above)
Comment (bertschneider) [reply src/main/java/com/code_intelligence/jazzer/junit/AgentConfigurator.java:40]: It shouldn't hurt, though. Ignore this for now.

File: src/main/java/com/code_intelligence/jazzer/driver/OptItem.java [hunk 5]
```diff
@@ -0,0 +1,383 @@
...
+
+  protected abstract String getType();
+
+  protected T accumulate(T oldValue, T newValue) {
+    return newValue;
+  }
+
+  private Optional<T> getExplicitValue() {
+    return Stream
+        .<Supplier<Stream<T>>>of(this::getFromManifest, this::getFromEnv, this::getFromProperties,
```
[the remark this answers is not in the document]
Comment (bertschneider) [reply src/main/java/com/code_intelligence/jazzer/driver/OptItem.java:167]: The rusulting code looks quite nice and more cleanly separated, it complicates CLI and JUnit configuration handling a little bit, though. Ignore it for now.

**Review by bertschneider — APPROVED** [review]

**Review by br-lewis — CHANGES_REQUESTED** [review]

File: src/main/java/com/code_intelligence/jazzer/driver/OptItem.java [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+    }
+
+    @Override
+    protected Optional<Boolean> fromString(String rawValue) throws IllegalOptionValueException {
+      return Optional.of(Boolean.valueOf(rawValue));
+    }
+
+    @Override
+    protected String getType() {
+      return null;
```
[the remark this answers is not in the document]
Comment (br-lewis) [reply src/main/java/com/code_intelligence/jazzer/driver/OptItem.java:272]: I'll just change it to return `bool` for now. We'll probably want to clean it up later and have `OptItem.Bool` override `toString` to drop everything in the parens but that can be done in another PR.

**Pull request merged**